### ISSUE TITLE
TypeError: Cannot read property 'module' of undefined

### ISSUE THREAD

**Issue by jordyvandomselaar** [storybooks/storybook#5877]
I'm using the latest Storybook: 5.0.0

I'm using React generated with `create-react-app` with the `--typescript` flag. Following this guide; https://storybook.js.org/docs/configurations/typescript-config/#setting-up-typescript-with-babel-loader

I get the following error: TypeError: Cannot read property 'module' of undefined

**Comment by tograd** [comment]
same issue I guess 

```
ERR! TypeError: Cannot read property 'module' of undefined
ERR!     at module.exports (D:\code\github\...\.storybook\webpack.config.js:3:10)
ERR!     at Object.webpack (D:\code\github\...\node_modules\@storybook\core\dist\server\preview\custom-webpack-preset.js:43:12)
ERR!  { TypeError: Cannot read property 'module' of undefined
ERR!     at module.exports (D:\code\github\...\.storybook\webpack.config.js:3:10)
ERR!     at Object.webpack (D:\code\github\...\node_modules\@storybook\core\dist\server\preview\custom-webpack-preset.js:43:12)
ERR!   stack:
ERR!    "TypeError: Cannot read property 'module' of undefined
    at module.exports (D:\\code\\github\\...\\.storybook\\webpack.config.js:3:10)
    at Object.webpack (D:\\code\\github\\...\\node_modules\\@storybook\\core\\dist\\server\\preview\\custom-webpack-preset.js:43:12)" }
```

**Comment by jangerhofer** [comment]
Logging the three variables in the line `(baseConfig, env, config) => {` shows that the latter two are `undefined` with v5.  The `baseConfig` object has a `config` key whose value is an object with `module` and `resolve`.  

I don't have any grasp on what the aforementioned trio of variables are intended to represent, but I did find that changing the enclosing function got me up and running.  My `webpack.config.js` now contains:

```
const path = require("path");
module.exports = ({ config }) => {
  config.module.rules.push({
    test: /\.(ts|tsx)$/,
    use: [
      {
        loader: require.resolve("awesome-typescript-loader")
      },
      {
        loader: require.resolve("react-docgen-typescript-loader")
      }
    ]
  });
  config.resolve.extensions.push(".ts", ".tsx");
  return config;
};
```

**Comment by tograd** [comment]
Thank you

Another issue that might crop up afterwards is in `.storybook/config.js`
```
import { configure } from '@storybook/react';
// automatically import all files ending in *.stories.tsx
const req = require.context('../stories', true, /.stories.tsx$/);

function loadStories() {
  req.keys().forEach(req);
}

configure(loadStories, module);
```
using the default npx install it puts files into `../src/stories` not `../stories`

**Comment by tmeasday** [comment]
Any one keen to send a PR to the docs to fix these two things? 🙏

**Comment by filiplindbladh** [comment]
Can't get the example with babel-loader to work either! tried some different webpack configs but no luck, it looks like this: 

```
const path = require("path");
module.exports = ({ config, mode }) => {
    config.module.rules.push({
        test: /\.(ts|tsx)$/,
        use: [{ loader: require.resolve("babel-loader") }],
    });
    config.resolve.extensions.push(".ts", ".tsx");
    return config;
};
```

any ideas?

**Comment by dai4** [comment]
@tograd Did you manage to find a workaround for the config issue?

**Comment by tograd** [comment]
@dai4 

yes just change to `require.context('../src', true, /.stories.tsx$/)` if you put `.stories` files everywhere, or `require.context('../src/stories', true, /.stories.tsx$/)` if you keep them in `src/stories` folder only

**Comment by dai4** [comment]
@tograd Thanks for your reply!

I misunderstood what you said as I'm having a different issue where this bit of code `require.context('../stories', true, /.stories.tsx$/)` is not actually looking only in the `../stories` folder, but rather everywhere, starting from root.

**Comment by jangerhofer** [comment]
@tmeasday: Sure!  I would be happy to take a stab at it.  It looks to me like [this line](https://github.com/storybooks/storybook/blob/2cf21e6f097c80e721f5a65a92974031887f9e41/lib/core/src/server/preview/custom-webpack-preset.js#L27) loads the function provided in a custom `webpack.config.js` file.  I see that said function's signature changed [in this commit](https://github.com/storybooks/storybook/commit/cb15153810e885ce90b0ba9782f8ef57fef4d627#diff-87ebcdfe91d2901199c5984f07014578R27).

I'll submit a PR for the docs accordingly!

**Comment by jangerhofer** [comment]
@filiplindbladh: Are you blocked by any error in particular?  I just tried using the `npx -p @storybook/cli sb init` set up method on a fresh Create-React-App Typescript project, and the `webpack.config.js` below (straight from [the docs](https://storybook.js.org/docs/configurations/typescript-config/#setting-up-typescript-to-work-with-storybook-1) worked on the first go for me.

```
module.exports = ({ config }) => {
  config.module.rules.push({
    test: /\.(ts|tsx)$/,
    loader: require.resolve("babel-loader"),
    options: {
      presets: [["react-app", { flow: false, typescript: true }]]
    }
  });
  config.resolve.extensions.push(".ts", ".tsx");
  return config;
};
```

**Comment by DanRYoung** [comment]
@jangerhofer For me at least, the issue arises in _upgrading_ Storybook from 4.x to 5 on my TypeScript CRA app, [per the upgrade guide](https://medium.com/storybookjs/storybook-5-migration-guide-d804b38c739d).

**Comment by littlepoolshark** [comment]
@jangerhofer 
It work for me,Thanks a lot.

**Comment by shilman** [comment]
@DanRYoung did you get this figured out? There was a bug in "extend-mode" webpack config that got fixed in 5.0.2 https://github.com/storybooks/storybook/blob/next/MIGRATION.md#deprecate-webpack-extend-mode

**Comment by jordyvandomselaar** [comment]
Thanks everyone for the awesome replies and all the help! You guys are awesome 👍

**Comment by forrestblade** [comment]
I'm still having this error, I'm not understanding what I need to do to fix it.  This is what my webpack.config inside my .storybook looks like. 
```
const path = require('path');
const fs = require('fs');
const appDirectory = fs.realpathSync(process.cwd());

module.exports = (baseConfig, env, defaultConfig) => {
    // Extend defaultConfig as you need.

    // For example, add storysource loader:
    defaultConfig.module.rules.push({
        test: /\.stories\.js?$/,
        loaders: [require.resolve('@storybook/addon-storysource/loader')],
        enforce: 'pre',
    });

    // Had to add this and the below packages in order for create-react-app
    // inline SVG loader to work.
    // https://github.com/storybooks/storybook/issues/4298
    defaultConfig.module.rules.push({
        test: /\.(js|jsx)$/,
        include: path.resolve(appDirectory, 'src'),
        loader: require.resolve('babel-loader'),
        options: {
            presets: [require.resolve('babel-preset-react-app')],
            plugins: [
                [
                    require.resolve('babel-plugin-named-asset-import'),
                    {
                        loaderMap: {
                            svg: {
                                ReactComponent: '@svgr/webpack?-prettier,-svgo![path]',
                            },
                        },
                    },
                ],
            ]
        }
    });

    defaultConfig.resolve.extensions.push(".stories.js");

    return defaultConfig;
};
```

**Comment by shilman** [comment]
@forrestblade If you're using Storybook 5, the webpack config function signature changed. You're still using the v4 version. See: https://github.com/storybooks/storybook/blob/next/MIGRATION.md#webpack-config-simplification

**Comment by forrestblade** [comment]
@shilman Sorry that config was not accurate, I updated to 5.0.11 and I changed the `module.exports = (baseConfig, env, defaultConfig) => {` to `module.exports = (config, mode) => {`

I get the same error except it's not on modules its on the rules.
` TypeError: Cannot read property 'rules' of undefined`

**Comment by forrestblade** [comment]
I updated my original comment to my current settings

**Comment by forrestblade** [comment]
I figured, it out.  It's been a long day haha. I forgot my curly braces

**Comment by sandinosaso** [comment]
For me to make it work with 5.3.x version I had to have my webpack config like this:

```
const path = require("path");
module.exports = ({ config }) => {
  return {
    ...config,
    module: {
      ...config.module,
      rules: [
        ...config.module.rules,
        {
          test: /\.(ts|tsx)$/,
          use: [
            {
              loader: require.resolve("awesome-typescript-loader")
            },
            {
              loader: require.resolve("react-docgen-typescript-loader")
            }
          ]
        }
      ]
    },
    resolve: {
      ...config.resolve,
      extensions: [
        ...config.resolve.extensions,
        ".ts",
        ".tsx"
      ]
    }
  }
};
```
